### PR TITLE
Speeding up docs build in CI using caching.

### PR DESCRIPTION
**Pull request by dpeachpeach**
FOR ISSUE #32687 

**Description**

This speeds up the 'Docs Build' step of the docs build job by caching the results of building the docs from prior actions. 

This kind of 'skips' the step 'build docs' altogether (in the successful case where there is a cache hit) and doesn't really provide any improvement to speeding up building the docs in case they were modified.

 I didn't know if that's what you guys were going for but I thought it could be as a decent amount of the contributions made don't affect building the docs directly. In those cases it speeds up the action by eliminating building the docs entirely. (Drawback here is another element present in the cache fields.)

I was also thinking of other approaches to speed things up. If this isn't sufficient / what you're looking for I can find a different approach and improve it further! Give me any feedback you have I'll go back to work on this.

**Verification**

I tested the action on my own fork, which showed that it cached correctly and provided an elimination of the 'build docs' step in case of cache hit. It might take two runs to store the cache properly but barring that it seems to be a success.

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
This is close but not quite right.

We want to build able to incrementally build using the cache, like we can with our scons build cache.

**Comment by dpeachpeach** [comment]
I took your feedback and tried to improve it further. I think I've got it pending your review but I just wanted to provide some explanation behind the edits + broad strokes bullet points:

- It turns out Sphinx already implements caching implicitly! The reason why it didn't manifest itself in building the docs for comma was that the build directory would be deleted on every build. That is where the automatically generated caches (or pickles) (.doctrees) were stored. I stopped that behavior and now it runs much faster.
- I also added in parallelism at the sphinx-build call, but when I had this undergo testing on my fork it didn't change anything about the performance (probably because parallelism is already done through the github action). This can be removed to clean up the code.
- I went through the route of looking at some optimizations that could affect the theme but I figured those were design decisions and I didn't want to intrude.
- I changed the paths of some of the directory alias definitions to more closely follow definitions I saw in the docs.

When undergoing testing on my fork, this seems to have resulted in the same results with faster compilation time. 
Let me know if you have any more feedback!

**Comment by adeebshihadeh** [comment]
Sounds good, I'll review once you mark it as ready for review.

> * I went through the route of looking at some optimizations that could affect the theme but I figured those were design decisions and I didn't want to intrude.

We basically got it setup and haven't touched it since, so design changes are welcome!

**Comment by adeebshihadeh** [comment]
I see a whole bunch of warnings and errors; are those not new?

https://github.com/commaai/openpilot/actions/runs/9589108324/job/26442355846?pr=32766

**Comment by dpeachpeach** [comment]
They seem to be the same warnings / errors that were present in the sample you showed in the issue:
https://github.com/commaai/openpilot/actions/runs/9454981147/job/26043750673

However I was worried about that. I was going to try to debug a little bit to see if any of those were related to my implementation before reviewing.

**Comment by dpeachpeach** [comment]
Just found a serious issue @adeebshihadeh, going to have to work on this a bit more. Sorry for prematurely setting up the PR!

**Comment by adeebshihadeh** [comment]
No worries, take your time :)

You can also share your new docs on Discord and ask people to poke around for bugs

**Comment by dpeachpeach** [comment]
All right, fixed the issue! When I test the generated html through local hosting, everything seems equivalent. (Compared it side by side with the docs available online). I'll make a post on the discord for anyone to take a look as well but otherwise I think this is ready!

File: docs/Makefile [hunk 1]
```diff
@@ -25,19 +25,20 @@ clean:
...
-	@echo "Cleaning build folder..."
-	rm -rf "$(BUILDDIR)"
+	# No longer removing the build each time.
+	# Sphinx natively supports caching through a .doctrees dir in the build dir
+
 	mkdir -p "$(DOCSBUILDDIR)"
 
 	@echo "Copying docs & config to build folder..."
-	cp -a "$(DOCSDIR)" "$(BUILDDIR)"
+	# cp -a "$(DOCSDIR)" "$(BUILDDIR)"
```
Comment (adeebshihadeh) [inline docs/Makefile:34]: this should either be added back or removed

**Review by adeebshihadeh — COMMENTED** [review]
It looks like this now fixes the docs build system to build incrementally, but I don't see how it caches in CI to take advantage of that?

File: docs/Makefile (hunk 1 above)
Comment (dpeachpeach) [reply docs/Makefile]: Removed it

**Comment by dpeachpeach** [comment]
> It looks like this now fixes the docs build system to build incrementally, but I don't see how it caches in CI to take advantage of that?

Initially, when I started this PR I thought the solution to speeding up 'build docs' lay somewhere in the github action. Once you mentioned incrementally building the cache I gave up speeding anything up through modifying the CI implementation and pivoted to trying to fix sphinx-build's behavior. (And why it wasn't retrieving its cache for incremental builds) 

It seems like fixing sphinx-build has enabled the 'build docs' step to reach the target time (even though I don't do any explicit caching in the CI) because sphinx automatically builds a cache in the form of the doctrees directory, which is present in the build directory that the github action outputs. Now that we're not removing the build directory on every iteration of running the action, we get the CI-level caching for free.

A lot of this is stuff I've learned in the past day or so, but I believe it to be correct. I'm no expert so please provide any feedback / correct me if I'm completely wrong.

**Pull request closed without merging**